File: src/Python/config/json_loader.py
from __future__ import annotations
import importlib
import json
from typing import Any, Callable, Dict, Optional

from ..core.pipeline import Pipeline, Pipe
from ..core.jumps import jump_now, jump_after
from ..core.metrics import Metrics, NoopMetrics

class PipelineJsonLoader:
    def __init__(self, *, beans: Optional[Dict[str, Any]]=None, instance: Optional[Any]=None, metrics: Optional[Metrics]=None):
        self.beans = dict(beans or {})
        self.instance = instance
        self.metrics = metrics or NoopMetrics()

    def load_str(self, s: str):
        spec = json.loads(s)
        return self._build(spec)

    def load_file(self, path: str):
        with open(path, 'r', encoding='utf-8') as f:
            return self._build(json.load(f))

    def _build(self, root: Dict[str, Any]):
        name = root.get('pipeline', 'pipeline')
        typ  = root.get('type', 'unary')
        sc   = bool(root.get('shortCircuit', True))

        if typ == 'unary':
            p = Pipeline(name, short_circuit=sc, metrics=self.metrics)
            for sec in ('pre', 'steps', 'post'):
                for node in root.get(sec, []) or []:
                    fn, label = self._compile_step(node)
                    section = 'main' if sec == 'steps' else sec
                    p.step(fn, label=label, section=section)
            return p

        elif typ == 'typed':
            pipe = Pipe(name, short_circuit=sc, metrics=self.metrics)
            for sec in ('pre', 'steps', 'post'):
                for node in root.get(sec, []) or []:
                    fn, _ = self._compile_step(node, typed=True)
                    pipe.step(fn)
            return pipe
        else:
            raise ValueError(f"Unsupported pipeline type: {typ}")

    def _compile_step(self, node: Dict[str, Any], typed: bool=False):
        label = node.get('label')

        jw = node.get('jumpWhen')
        pred_fn = None
        jump_label = None
        jump_delay = 0
        if jw:
            jump_label = jw.get('label')
            jump_delay = int(jw.get('delayMillis', 0))
            pred_spec = jw.get('predicate')
            if pred_spec:
                pred_fn, _ = self._compile_step(pred_spec, typed=True)

        if '$local' in node:
            fqcn = node['$local']
            step_callable = self._resolve_local(fqcn)

        elif '$method' in node:
            step_callable = self._resolve_method(node['$method'])

        elif '$prompt' in node:
            step_callable = self._resolve_prompt(node['$prompt'])

        elif '$remote' in node:
            step_callable = self._resolve_remote(node['$remote'])

        else:
            raise ValueError(f"Unsupported step node: {node}")

        if jw and pred_fn and jump_label:
            def wrapped(x, _pred=pred_fn, _inner=step_callable, _label=jump_label, _delay=jump_delay):
                if bool(_pred(x)):
                    if _delay > 0: jump_after(_label, _delay)
                    else: jump_now(_label)
                return _inner(x)
            return wrapped, label
        else:
            return step_callable, label

    def _resolve_local(self, fqcn: str):
        module_name, cls_name = fqcn.rsplit('.', 1)
        mod = importlib.import_module(module_name)
        cls = getattr(mod, cls_name)
        inst = cls()
        if not hasattr(inst, 'apply'):
            raise TypeError(f"Class {fqcn} lacks an 'apply' method")
        return getattr(inst, 'apply')

    def _resolve_method(self, spec: Dict[str, Any]):
        ref = spec.get('ref')
        target = spec.get('target')
        if '#' in ref:
            mod_cls, meth = ref.split('#', 1)
            module_name, cls_name = mod_cls.rsplit('.', 1)
            mod = importlib.import_module(module_name)
            cls = getattr(mod, cls_name)
            if target:
                if target == '@this':
                    obj = self.instance
                    if obj is None: raise ValueError("target=@this but loader 'instance' is None")
                elif target.startswith('@'):
                    bean_id = target[1:]
                    if bean_id not in self.beans: raise ValueError(f"Unknown bean id: {bean_id}")
                    obj = self.beans[bean_id]
                else:
                    raise ValueError(f"Unsupported target: {target}")
                return getattr(obj, meth)
            else:
                return getattr(cls, meth)
        else:
            module_name, func_name = ref.rsplit(':', 1)
            mod = importlib.import_module(module_name)
            return getattr(mod, func_name)

    def _resolve_prompt(self, spec: Dict[str, Any]):
        adapter = self.beans.get('llm_adapter')
        if adapter is None:
            def _raise(_): raise NotImplementedError("No 'llm_adapter' bean provided for $prompt step")
            return _raise
        def _call(x, _adapter=adapter, _spec=spec):
            return _adapter(x, _spec)
        return _call

    def _resolve_remote(self, spec: Dict[str, Any]):
        from ..remote.http_step import http_step, RemoteSpec
        rs = RemoteSpec(
            endpoint = spec['endpoint'],
            timeout_millis = int(spec.get('timeoutMillis', 1000)),
            retries = int(spec.get('retries', 0)),
            headers = dict(spec.get('headers', {})),
            method = spec.get('method', 'POST').upper(),
        )
        to_json_id = spec.get('toJsonBean')
        from_json_id = spec.get('fromJsonBean')
        rs.to_json = self.beans[to_json_id] if to_json_id else (lambda i: json.dumps(i))
        rs.from_json = self.beans[from_json_id] if from_json_id else (lambda s: json.loads(s))
        return http_step(rs)

File: src/typescript/config/json_loader.ts
import { Pipeline, Pipe, StepFunction } from "../core/pipeline.js";
import { jumpAfter, jumpNow } from "../core/jumps.js";
import { Metrics, NoopMetrics } from "../core/metrics.js";

export interface LoaderOptions {
  beans?: Record<string, unknown>;
  instance?: unknown;
  metrics?: Metrics;
}

export class PipelineJsonLoader {
  private beans: Record<string, unknown>;
  private instance: unknown;
  private metrics: Metrics;

  constructor(options?: LoaderOptions) {
    this.beans = { ...(options?.beans ?? {}) };
    this.instance = options?.instance;
    this.metrics = options?.metrics ?? new NoopMetrics();
  }

  async loadStr(jsonText: string): Promise<unknown> {
    const spec = JSON.parse(jsonText);
    return this.build(spec);
  }

  async loadFile(filePath: string): Promise<unknown> {
    const fs = await import("node:fs/promises");
    const text = await fs.readFile(filePath, { encoding: "utf-8" });
    return this.loadStr(text);
  }

  async build(root: any): Promise<unknown> {
    const pipelineName = String(root.pipeline ?? "pipeline");
    const pipelineType = String(root.type ?? "unary");
    const shortFlag = Boolean(root.shortCircuit ?? true);

    if (pipelineType === "unary") {
      const pipeline = new Pipeline<any>(pipelineName, shortFlag, this.metrics);
      for (const sectionName of ["pre", "steps", "post"] as const) {
        const nodes: any[] = root[sectionName] ?? [];
        for (const node of nodes) {
          const { step, label } = await this.compileStep(node, false);
          const section = sectionName === "steps" ? "main" : (sectionName as "pre" | "post");
          pipeline.step(step, { label, section });
        }
      }
      return pipeline;
    } else if (pipelineType === "typed") {
      const pipe = new Pipe<any, any>(pipelineName, shortFlag, this.metrics);
      for (const sectionName of ["pre", "steps", "post"] as const) {
        const nodes: any[] = root[sectionName] ?? [];
        for (const node of nodes) {
          const { step } = await this.compileStep(node, true);
          pipe.step(step as any);
        }
      }
      return pipe;
    } else {
      throw new Error(`Unsupported pipeline type: ${pipelineType}`);
    }
  }

  private async compileStep(node: any, typed: boolean): Promise<{ step: StepFunction<any>; label?: string }> {
    const label = node.label as string | undefined;

    const jumpWhen = node.jumpWhen as any | undefined;
    let predicateStep: StepFunction<any> | undefined;
    let jumpLabel: string | undefined;
    let jumpDelayMs: number = 0;

    if (jumpWhen) {
      jumpLabel = String(jumpWhen.label ?? "");
      jumpDelayMs = Number(jumpWhen.delayMillis ?? 0);
      if (jumpWhen.predicate) {
        const compiled = await this.compileStep(jumpWhen.predicate, true);
        predicateStep = compiled.step;
      }
    }

    let stepCallable: StepFunction<any>;

    if (node["$local"]) {
      const fqn = String(node["$local"]);
      stepCallable = await this.resolveLocal(fqn);
    } else if (node["$method"]) {
      stepCallable = await this.resolveMethod(node["$method"]);
    } else if (node["$prompt"]) {
      stepCallable = this.resolvePrompt(node["$prompt"]);
    } else if (node["$remote"]) {
      stepCallable = this.resolveRemote(node["$remote"]);
    } else {
      throw new Error("Unsupported step node");
    }

    if (jumpWhen && predicateStep && jumpLabel) {
      const wrapped: StepFunction<any> = async (inputValue: any) => {
        const truthy = Boolean(await predicateStep!(inputValue));
        if (truthy) {
          if (jumpDelayMs > 0) {
            jumpAfter(jumpLabel!, jumpDelayMs);
          } else {
            jumpNow(jumpLabel!);
          }
        }
        const result = await stepCallable(inputValue);
        return result;
      };
      return { step: wrapped, label };
    } else {
      return { step: stepCallable, label };
    }
  }

  private async resolveLocal(fqn: string): Promise<StepFunction<any>> {
    // Interpret "package.module.ClassName" as "package/module" and "ClassName"
    const lastDot = fqn.lastIndexOf(".");
    if (lastDot < 0) {
      throw new Error(`Invalid $local value (expected module.Class): ${fqn}`);
    }
    const moduleName = fqn.slice(0, lastDot).replaceAll(".", "/");
    const className = fqn.slice(lastDot + 1);
    const moduleObj: any = await import(moduleName);
    const classRef = moduleObj[className];
    if (!classRef) {
      throw new Error(`Class ${className} not found in module ${moduleName}`);
    }
    const instance = new classRef();
    if (typeof instance.apply !== "function") {
      throw new Error(`Class ${fqn} lacks an 'apply' method`);
    }
    const callApply: StepFunction<any> = async (inputValue: any) => {
      return await instance.apply(inputValue);
    };
    return callApply;
  }

  private async resolveMethod(spec: any): Promise<StepFunction<any>> {
    const ref = String(spec.ref ?? "");
    const target = spec.target ? String(spec.target) : "";

    if (ref.includes("#")) {
      const [moduleClass, methodName] = ref.split("#");
      const lastDot = moduleClass.lastIndexOf(".");
      if (lastDot < 0) {
        throw new Error(`Invalid method ref: ${ref}`);
      }
      const moduleName = moduleClass.slice(0, lastDot).replaceAll(".", "/");
      const className = moduleClass.slice(lastDot + 1);
      const moduleObj: any = await import(moduleName);
      const classRef = moduleObj[className];
      if (!classRef) {
        throw new Error(`Class ${className} not found in module ${moduleName}`);
      }

      let bound: any;
      if (target === "@this") {
        bound = this.instance;
        if (bound == null) {
          throw new Error("target=@this but loader 'instance' is undefined");
        }
      } else if (target.startsWith("@")) {
        const beanId = target.slice(1);
        bound = this.beans[beanId];
        if (bound == null) {
          throw new Error(`Unknown bean id: ${beanId}`);
        }
      } else {
        bound = classRef;
      }

      const method = bound[methodName];
      if (typeof method !== "function") {
        throw new Error(`Method ${methodName} not found on ${target || className}`);
      }
      const callMethod: StepFunction<any> = async (inputValue: any) => {
        return await method.call(bound, inputValue);
      };
      return callMethod;
    } else {
      const lastColon = ref.lastIndexOf(":");
      if (lastColon < 0) {
        throw new Error(`Invalid function ref: ${ref}`);
      }
      const moduleName = ref.slice(0, lastColon).replaceAll(".", "/");
      const funcName = ref.slice(lastColon + 1);
      const moduleObj: any = await import(moduleName);
      const func = moduleObj[funcName];
      if (typeof func !== "function") {
        throw new Error(`Function ${funcName} not found in module ${moduleName}`);
      }
      const callFunc: StepFunction<any> = async (inputValue: any) => {
        return await func(inputValue);
      };
      return callFunc;
    }
  }

  private resolvePrompt(promptSpec: any): StepFunction<any> {
    const adapter = this.beans["llm_adapter"];
    if (typeof adapter !== "function") {
      const raise: StepFunction<any> = async () => {
        throw new Error("No 'llm_adapter' bean provided for $prompt step");
      };
      return raise;
    }
    const callAdapter: StepFunction<any> = async (inputValue: any) => {
      return await (adapter as Function)(inputValue, promptSpec);
    };
    return callAdapter;
  }

  private resolveRemote(remoteSpec: any): StepFunction<any> {
    const { httpStep, RemoteSpec } = requireRemote();
    const spec = new RemoteSpec({
      endpoint: String(remoteSpec.endpoint),
      timeoutMillis: Number(remoteSpec.timeoutMillis ?? 1000),
      retries: Number(remoteSpec.retries ?? 0),
      headers: { ...(remoteSpec.headers ?? {}) },
      method: String(remoteSpec.method ?? "POST").toUpperCase()
    });
    const toJsonId = remoteSpec.toJsonBean;
    const fromJsonId = remoteSpec.fromJsonBean;
    if (toJsonId && typeof this.beans[toJsonId] === "function") {
      spec.toJson = this.beans[toJsonId] as (inputValue: unknown) => string;
    }
    if (fromJsonId && typeof this.beans[fromJsonId] === "function") {
      spec.fromJson = this.beans[fromJsonId] as (text: string) => unknown;
    }
    return httpStep(spec);
  }
}

function requireRemote() {
  // Local import to avoid circular import issues in ESM
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const mod = require("../remote/http_step.js");
  return mod;
}
